native/src/main/native/org/mozilla/jss/pkcs11/PK11KeyGenerator.c

```diff
@@ -183,6 +183,73 @@ print_secitem(SECItem *item) {
     }
 }
 
+
+/***********************************************************************
+ *
+ * c o n s t r u c t S H A P B E K e y
+ *
+ * Constructs a PBE key using CKM_SHA*.  This should
+ * be supported by NSS automatically, but isn't.
+ *
+ * RETURNS
+ *      A symmetric key from the given password, salt, and iteration count,
+ *      or NULL if an exception was thrown.
+ * THROWS
+ *      TokenException if an error occurs.
+ */
+static PK11SymKey*
+constructSHAPBEKey(JNIEnv *env, CK_MECHANISM_TYPE mech, PK11SlotInfo *slot, SECItem *pwitem, SECItem *salt,
+        int iterationCount)
+{
+    PK11SymKey *key=NULL;
+    SECItem *params=NULL;
+
+    if( pwitem == NULL ) {
+        JSS_throwMsg(env, TOKEN_EXCEPTION,
+            "constructSHAPBEKey:"
+            " pwitem NULL");
+        goto finish;
+    }
+    if( salt == NULL ) {
+        JSS_throwMsg(env, TOKEN_EXCEPTION,
+            "constructSHAPBEKey:"
+            " salt NULL");
+        goto finish;
+    }
+
+    params = PK11_CreatePBEParams(salt, pwitem,
+        iterationCount);
+    switch(mech){
```

Comment:
Looks like maybe we have some tab / space formatting issues here with respect to the switch clause and the lines after. Minor if you get around to it.

Reply:
I cannot find a good way to correctly format java and C in eclipse. Fixed using vim.